packages/core/index.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * @typedef {'gpt-4' | 'gpt-4-0314' | 'gpt-4-32k' | 'gpt-4-32k-0314' | 'gpt-3.5-turbo' | 'gpt-3.5-turbo-0301'} OpenAIChatCompletionsModelId
+ * @typedef {'text-davinci-003' | 'text-davinci-002' | 'text-curie-001' | 'text-babbage-001' | 'text-ada-001' | 'davinci' | 'curie' | 'babbage' | 'ada'} OpenAICompletionsModelId
+ * @typedef {OpenAIChatCompletionsModelId | OpenAICompletionsModelId} OpenAIModelId
+ */
+
+/**
+ * @type {OpenAIModelId}
+ */
+export const DEFAULT_MODEL = 'gpt-3.5-turbo';
+export const I_DONT_KNOW_MESSAGE = 'Sorry, I am not sure how to answer that.';
+export const MARKPROMPT_COMPLETIONS_URL =
+  'https://api.markprompt.com/v1/completions';
+export const STREAM_SEPARATOR = '___START_RESPONSE_STREAM___';
+
+const defaultOptions = {
+  model: DEFAULT_MODEL,
+  completionsUrl: MARKPROMPT_COMPLETIONS_URL,
+  iDontKnowMessage: I_DONT_KNOW_MESSAGE,
+  promptTemplate: undefined,
+};
+
+/**
+ * @typedef {Object} Options
+ * @property {string} [completionsUrl] - URL at which to fetch completions
+ * @property {string} [iDontKnowMessage] - Message returned when the model does not have an answer
+ * @property {OpenAIModelId} [model] - The model to use
+ * @property {string} [promptTemplate] - The prompt template
+ */
+
+/**
+ * @param {string} prompt - Prompt to submit to the model
+ * @param {string} projectKey - The key of your project
+ * @param {(answerChunk: string) => void} onAnswerChunk - Answers come in via streaming. This function is called when a new chunk arrives
+ * @param {(references: string[]) => void} onReferences - This function is called when a chunk includes references.
+ * @param {Options} [options] - Optional options object
+ */
+export const submitPrompt = async function (
+  prompt,
+  projectKey,
+  onAnswerChunk,
+  onReferences,
+  options,
+) {
+  if (!projectKey) {
+    throw new Error('A projectKey is required.');
+  }
+
+  if (!prompt) return;
+
+  options = Object.fromEntries(
+    Object.entries(defaultOptions).map(([key, value]) => [
+      key,
+      options?.[key] ?? value,
+    ]),
+  );
```

Comment:
I think a simple spread would work here as well.

```suggestion
  options = {
    ...defaultOptions,
    ...options
  };
```

It may also be better to not reassign to function parameters.

Reply:
No, unfortunately won't work. When we consume this in `@markprompt/react` some values will be passed from optional props. That means that in the `options` object they can have `undefined` as a value. This makes sure that options that are defined but have `undefined` as a value will take the default value.

Reply:
Actually, it might may be even better to remove this statement, and below use:

```ts
JSON.stringify({
  prompt,
  model: options?.model ?? DEFAULT_MODEL,
  completionsUrl: options?.completionsUrl ?? MARKPROMPT_COMPLETIONS_URL,
  iDontKnowMessage: options?.iDontKnowMessage ?? I_DONT_KNOW_MESSAGE,
  promptTemplate: options?.promptTemplate,
  projectKey,
})
```